File: api/app/schemas/document.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class DocumentChapterPublic(BaseModel):
    id: int
    name: str
    description: str | None = None
    parent_id: int | None = None
    sort_order: int
    created_at: datetime
    updated_at: datetime


class DocumentChapterTreeItem(DocumentChapterPublic):
    children: list["DocumentChapterTreeItem"] = Field(default_factory=list)
    documents: list["DocumentPublic"] = Field(default_factory=list)


class DocumentChapterListResponse(BaseModel):
    items: list[DocumentChapterPublic]
    total: int


class DocumentChapterCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=512)
    parent_id: int | None = None
    sort_order: int = 0


class DocumentChapterUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=512)
    parent_id: int | None = None
    sort_order: int | None = None


class DocumentPublic(BaseModel):
    id: int
    title: str
    content: str
    chapter_id: int | None = None
    sort_order: int
    status: str
    created_at: datetime
    updated_at: datetime


class DocumentListResponse(BaseModel):
    items: list[DocumentPublic]
    total: int


class DocumentCreateRequest(BaseModel):
    title: str = Field(min_length=1, max_length=256)
    content: str
    chapter_id: int | None = None
    sort_order: int = 0
    status: Literal["draft", "published"] = "draft"


class DocumentUpdateRequest(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=256)
    content: str | None = None
    chapter_id: int | None = None
    sort_order: int | None = None
    status: Literal["draft", "published"] | None = None

File: web/src/types/document.ts
export interface DocumentChapter {
  id: number;
  name: string;
  description: string | null;
  parent_id: number | null;
  sort_order: number;
  created_at: string;
  updated_at: string;
}

export interface DocumentChapterTreeItem extends DocumentChapter {
  children: DocumentChapterTreeItem[];
  documents: Document[];
}

export interface DocumentChapterListResponse {
  items: DocumentChapter[];
  total: number;
}

export interface DocumentChapterCreateRequest {
  name: string;
  description?: string | null;
  parent_id?: number | null;
  sort_order?: number;
}

export interface DocumentChapterUpdateRequest {
  name?: string;
  description?: string | null;
  parent_id?: number | null;
  sort_order?: number;
}

export interface Document {
  id: number;
  title: string;
  content: string;
  chapter_id: number | null;
  sort_order: number;
  status: "draft" | "published";
  created_at: string;
  updated_at: string;
}

export interface DocumentListResponse {
  items: Document[];
  total: number;
}

export interface DocumentCreateRequest {
  title: string;
  content: string;
  chapter_id?: number | null;
  sort_order?: number;
  status?: "draft" | "published";
}

export interface DocumentUpdateRequest {
  title?: string;
  content?: string;
  chapter_id?: number | null;
  sort_order?: number;
  status?: "draft" | "published";
}
